src/addr-util.c

```diff
@@ -37,24 +37,31 @@ CattaAddress *catta_address_from_sockaddr(const struct sockaddr* sa, CattaAddres
     assert(sa->sa_family == AF_INET || sa->sa_family == AF_INET6);
 
     ret_addr->proto = catta_af_to_proto(sa->sa_family);
-
-    if (sa->sa_family == AF_INET)
-        memcpy(&ret_addr->data.ipv4, &((const struct sockaddr_in*) sa)->sin_addr, sizeof(ret_addr->data.ipv4));
-    else
-        memcpy(&ret_addr->data.ipv6, &((const struct sockaddr_in6*) sa)->sin6_addr, sizeof(ret_addr->data.ipv6));
-
+    if(sa->sa_family == AF_INET) {
+      struct sockaddr_in sa_in;
+      memcpy(&sa_in, sa, sizeof(struct sockaddr));
+      memcpy(&ret_addr->data.ipv4, &(&sa_in)->sin_addr, sizeof(ret_addr->data.ipv4));
+    } else {
+      struct sockaddr_in6 sa_in6;
+      memcpy(&sa_in6, sa, sizeof(struct sockaddr));
+      memcpy(&ret_addr->data.ipv6, &(&sa_in6)->sin6_addr, sizeof(ret_addr->data.ipv6));
+    }
     return ret_addr;
 }
 
 uint16_t catta_port_from_sockaddr(const struct sockaddr* sa) {
     assert(sa);
 
     assert(sa->sa_family == AF_INET || sa->sa_family == AF_INET6);
-
-    if (sa->sa_family == AF_INET)
-        return ntohs(((const struct sockaddr_in*) sa)->sin_port);
-    else
-        return ntohs(((const struct sockaddr_in6*) sa)->sin6_port);
+    if(sa->sa_family == AF_INET) {
+      struct sockaddr_in sa_in;
+      memcpy(&sa_in, sa, sizeof(struct sockaddr));
+      return ntohs((&sa_in)->sin_port);
```

Comment:
just use the ```.``` dot operator, instead of converting to a pointer and then using ```->``` operator.

So in this case: ```return ntohs(sa_in.sin_port)```